### 02. Analysis Seoul Crime

1. 프로젝트 개요

2. 데이터 개요

In [12]:
import numpy as np
import pandas as pd

In [13]:
# 데이터 읽기

crime_raw_data = pd.read_csv('../data/02. crime_in_Seoul.csv', thousands="," , encoding='euc-kr')
crime_raw_data.head()

구분  죄종 발생검거     건수
0  중부  살인   발생    2.0
1  중부  살인   검거    2.0
2  중부  강도   발생    3.0
3  중부  강도   검거    3.0
4  중부  강간   발생  141.0

In [14]:
crime_raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65534 entries, 0 to 65533
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   구분      310 non-null    object 
 1   죄종      310 non-null    object 
 2   발생검거    310 non-null    object 
 3   건수      310 non-null    float64
dtypes: float64(1), object(3)
memory usage: 2.0+ MB


- info() : 데이터의 개요 확인하기
- rangeindex가 65534 인데, 310개로 나온다. (문제점)

In [15]:
crime_raw_data['죄종'].unique()

array(['살인', '강도', '강간', '절도', '폭력', nan], dtype=object)

- 특정 컬럼에서 unique 값을 조사.
- nan 값이 들어가 있다.

In [16]:
crime_raw_data[crime_raw_data['죄종'].isnull()].head()

구분   죄종 발생검거  건수
310  NaN  NaN  NaN NaN
311  NaN  NaN  NaN NaN
312  NaN  NaN  NaN NaN
313  NaN  NaN  NaN NaN
314  NaN  NaN  NaN NaN

In [17]:
crime_raw_data = crime_raw_data[crime_raw_data['죄종'].notnull()]

In [18]:
crime_raw_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 310 entries, 0 to 309
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   구분      310 non-null    object 
 1   죄종      310 non-null    object 
 2   발생검거    310 non-null    object 
 3   건수      310 non-null    float64
dtypes: float64(1), object(3)
memory usage: 12.1+ KB


In [19]:
crime_raw_data.head()

구분  죄종 발생검거     건수
0  중부  살인   발생    2.0
1  중부  살인   검거    2.0
2  중부  강도   발생    3.0
3  중부  강도   검거    3.0
4  중부  강간   발생  141.0

In [20]:
crime_raw_data.tail()

구분  죄종 발생검거      건수
305  수서  강간   검거   144.0
306  수서  절도   발생  1149.0
307  수서  절도   검거   789.0
308  수서  폭력   발생  1666.0
309  수서  폭력   검거  1431.0

------------------

## pandas pivot table

- index, columns, values, aggfunc

In [21]:
df = pd.read_excel('../data/02. sales-funnel.xlsx')
df.head()

Account                          Name            Rep       Manager  \
0   714466               Trantow-Barrows   Craig Booker  Debra Henley   
1   714466               Trantow-Barrows   Craig Booker  Debra Henley   
2   714466               Trantow-Barrows   Craig Booker  Debra Henley   
3   737550  Fritsch, Russel and Anderson   Craig Booker  Debra Henley   
4   146832                  Kiehn-Spinka  Daniel Hilton  Debra Henley   

       Product  Quantity  Price     Status  
0          CPU         1  30000  presented  
1     Software         1  10000  presented  
2  Maintenance         2   5000    pending  
3          CPU         1  35000   declined  
4          CPU         2  65000        won

#### index 설정

In [22]:
# Name 컬럼을 인덱스로 설정

#pd.pivot_table(df, index="Name")

df.pivot_table(index="Name", values=['Account','Quantity', 'Price'])

Account   Price  Quantity
Name                                                   
Barton LLC                     740150   35000  1.000000
Fritsch, Russel and Anderson   737550   35000  1.000000
Herman LLC                     141962   65000  2.000000
Jerde-Hilpert                  412290    5000  2.000000
Kassulke, Ondricka and Metz    307599    7000  3.000000
Keeling LLC                    688981  100000  5.000000
Kiehn-Spinka                   146832   65000  2.000000
Koepp Ltd                      729833   35000  2.000000
Kulas Inc                      218895   25000  1.500000
Purdy-Kunde                    163416   30000  1.000000
Stokes LLC                     239344    7500  1.000000
Trantow-Barrows                714466   15000  1.333333

In [23]:
#멀티 인덱스 설정

df.pivot_table(index=['Name','Rep','Manager'], values=['Account','Quantity', 'Price'])

Account   Price  \
Name                         Rep           Manager                          
Barton LLC                   John Smith    Debra Henley    740150   35000   
Fritsch, Russel and Anderson Craig Booker  Debra Henley    737550   35000   
Herman LLC                   Cedric Moss   Fred Anderson   141962   65000   
Jerde-Hilpert                John Smith    Debra Henley    412290    5000   
Kassulke, Ondricka and Metz  Wendy Yule    Fred Anderson   307599    7000   
Keeling LLC                  Wendy Yule    Fred Anderson   688981  100000   
Kiehn-Spinka                 Daniel Hilton Debra Henley    146832   65000   
Koepp Ltd                    Wendy Yule    Fred Anderson   729833   35000   
Kulas Inc                    Daniel Hilton Debra Henley    218895   25000   
Purdy-Kunde                  Cedric Moss   Fred Anderson   163416   30000   
Stokes LLC                   Cedric Moss   Fred Anderson   239344    7500   
Trantow-Barrows              Craig Booker  Debra Henley    714466   15000   

                                                          Quantity  
Name                         Rep           Manager                  
Barton LLC                   John Smith    Debra Henley   1.000000  
Fritsch, Russel and Anderson Craig Booker  Debra Henley   1.000000  
Herman LLC                   Cedric Moss   Fred Anderson  2.000000  
Jerde-Hilpert                John Smith    Debra Henley   2.000000  
Kassulke, Ondricka and Metz  Wendy Yule    Fred Anderson  3.000000  
Keeling LLC                  Wendy Yule    Fred Anderson  5.000000  
Kiehn-Spinka                 Daniel Hilton Debra Henley   2.000000  
Koepp Ltd                    Wendy Yule    Fred Anderson  2.000000  
Kulas Inc                    Daniel Hilton Debra Henley   1.500000  
Purdy-Kunde                  Cedric Moss   Fred Anderson  1.000000  
Stokes LLC                   Cedric Moss   Fred Anderson  1.000000  
Trantow-Barrows              Craig Booker  Debra Henley   1.333333

#### values 설정

In [24]:
df.head()

Account                          Name            Rep       Manager  \
0   714466               Trantow-Barrows   Craig Booker  Debra Henley   
1   714466               Trantow-Barrows   Craig Booker  Debra Henley   
2   714466               Trantow-Barrows   Craig Booker  Debra Henley   
3   737550  Fritsch, Russel and Anderson   Craig Booker  Debra Henley   
4   146832                  Kiehn-Spinka  Daniel Hilton  Debra Henley   

       Product  Quantity  Price     Status  
0          CPU         1  30000  presented  
1     Software         1  10000  presented  
2  Maintenance         2   5000    pending  
3          CPU         1  35000   declined  
4          CPU         2  65000        won

In [25]:
df.pivot_table(index=['Manager', 'Rep'], values='Price')

Price
Manager       Rep                        
Debra Henley  Craig Booker   20000.000000
              Daniel Hilton  38333.333333
              John Smith     20000.000000
Fred Anderson Cedric Moss    27500.000000
              Wendy Yule     44250.000000

In [26]:
# price 컬럼 sum 연산 적용

df.pivot_table(index=['Manager', 'Rep'], values='Price', aggfunc=np.sum)

Price
Manager       Rep                  
Debra Henley  Craig Booker    80000
              Daniel Hilton  115000
              John Smith      40000
Fred Anderson Cedric Moss    110000
              Wendy Yule     177000

In [27]:
# price 컬럼 sum 연산 적용

df.pivot_table(index=['Manager', 'Rep'], values='Price', aggfunc=[np.sum, len])

sum   len
                              Price Price
Manager       Rep                        
Debra Henley  Craig Booker    80000     4
              Daniel Hilton  115000     3
              John Smith      40000     2
Fred Anderson Cedric Moss    110000     4
              Wendy Yule     177000     4

In [28]:
df.head()

Account                          Name            Rep       Manager  \
0   714466               Trantow-Barrows   Craig Booker  Debra Henley   
1   714466               Trantow-Barrows   Craig Booker  Debra Henley   
2   714466               Trantow-Barrows   Craig Booker  Debra Henley   
3   737550  Fritsch, Russel and Anderson   Craig Booker  Debra Henley   
4   146832                  Kiehn-Spinka  Daniel Hilton  Debra Henley   

       Product  Quantity  Price     Status  
0          CPU         1  30000  presented  
1     Software         1  10000  presented  
2  Maintenance         2   5000    pending  
3          CPU         1  35000   declined  
4          CPU         2  65000        won

In [29]:
# product 를 컬럼으로 지정
df.pivot_table(index=['Manager','Rep'], values="Price", columns='Product', aggfunc=np.sum)

Product                           CPU  Maintenance  Monitor  Software
Manager       Rep                                                    
Debra Henley  Craig Booker    65000.0       5000.0      NaN   10000.0
              Daniel Hilton  105000.0          NaN      NaN   10000.0
              John Smith      35000.0       5000.0      NaN       NaN
Fred Anderson Cedric Moss     95000.0       5000.0      NaN   10000.0
              Wendy Yule     165000.0       7000.0   5000.0       NaN

In [30]:
# Nan 값 설정 : fill_value
df.pivot_table(index=['Manager', 'Rep'], values='Price', columns='Product', aggfunc=np.sum, fill_value=0)

Product                         CPU  Maintenance  Monitor  Software
Manager       Rep                                                  
Debra Henley  Craig Booker    65000         5000        0     10000
              Daniel Hilton  105000            0        0     10000
              John Smith      35000         5000        0         0
Fred Anderson Cedric Moss     95000         5000        0     10000
              Wendy Yule     165000         7000     5000         0

In [31]:
# 2개 이상 index, values 설정
df.pivot_table(index=['Manager','Rep','Product'], values=['Price', 'Quantity'], aggfunc=np.sum, fill_value=0)

Price  Quantity
Manager       Rep           Product                      
Debra Henley  Craig Booker  CPU           65000         2
                            Maintenance    5000         2
                            Software      10000         1
              Daniel Hilton CPU          105000         4
                            Software      10000         1
              John Smith    CPU           35000         1
                            Maintenance    5000         2
Fred Anderson Cedric Moss   CPU           95000         3
                            Maintenance    5000         1
                            Software      10000         1
              Wendy Yule    CPU          165000         7
                            Maintenance    7000         3
                            Monitor        5000         2

In [32]:
# aggfunc 2개 이상 설정

df.pivot_table(
    index=['Manager','Rep','Product'],
    values=['Price', 'Quantity'],
    aggfunc=[np.sum,np.mean],
    margins=True, #총계 ALL 추가
    fill_value=0)

sum                   mean  \
                                          Price Quantity         Price   
Manager       Rep           Product                                      
Debra Henley  Craig Booker  CPU           65000        2  32500.000000   
                            Maintenance    5000        2   5000.000000   
                            Software      10000        1  10000.000000   
              Daniel Hilton CPU          105000        4  52500.000000   
                            Software      10000        1  10000.000000   
              John Smith    CPU           35000        1  35000.000000   
                            Maintenance    5000        2   5000.000000   
Fred Anderson Cedric Moss   CPU           95000        3  47500.000000   
                            Maintenance    5000        1   5000.000000   
                            Software      10000        1  10000.000000   
              Wendy Yule    CPU          165000        7  82500.000000   
                            Maintenance    7000        3   7000.000000   
                            Monitor        5000        2   5000.000000   
All                                      522000       30  30705.882353   

                                                   
                                         Quantity  
Manager       Rep           Product                
Debra Henley  Craig Booker  CPU          1.000000  
                            Maintenance  2.000000  
                            Software     1.000000  
              Daniel Hilton CPU          2.000000  
                            Software     1.000000  
              John Smith    CPU          1.000000  
                            Maintenance  2.000000  
Fred Anderson Cedric Moss   CPU          1.500000  
                            Maintenance  1.000000  
                            Software     1.000000  
              Wendy Yule    CPU          3.500000  
                            Maintenance  3.000000  
                            Monitor      2.000000  
All                                      1.764706

--------------------------------------

### 3. 서울시 범죄 현황 데이터 정리

In [33]:
crime_raw_data.head()

구분  죄종 발생검거     건수
0  중부  살인   발생    2.0
1  중부  살인   검거    2.0
2  중부  강도   발생    3.0
3  중부  강도   검거    3.0
4  중부  강간   발생  141.0

In [34]:
crime_station = crime_raw_data.pivot_table(
    crime_raw_data,
    index='구분',
    columns=['죄종', '발생검거'],
    aggfunc=[np.sum])

crime_station.head()

sum                                                              
         건수                                                              
죄종       강간           강도          살인           절도              폭력        
발생검거     검거     발생    검거    발생    검거   발생      검거      발생      검거      발생
구분                                                                       
강남    269.0  339.0  26.0  24.0   3.0  3.0  1129.0  2438.0  2096.0  2336.0
강동    152.0  160.0  13.0  14.0   5.0  4.0   902.0  1754.0  2201.0  2530.0
강북    159.0  217.0   4.0   5.0   6.0  7.0   672.0  1222.0  2482.0  2778.0
강서    239.0  275.0  10.0  10.0  10.0  9.0  1070.0  1952.0  2768.0  3204.0
관악    264.0  322.0  10.0  12.0   7.0  6.0   937.0  2103.0  2707.0  3235.0

In [35]:
crime_station.columns #multiindex

MultiIndex([('sum', '건수', '강간', '검거'),
            ('sum', '건수', '강간', '발생'),
            ('sum', '건수', '강도', '검거'),
            ('sum', '건수', '강도', '발생'),
            ('sum', '건수', '살인', '검거'),
            ('sum', '건수', '살인', '발생'),
            ('sum', '건수', '절도', '검거'),
            ('sum', '건수', '절도', '발생'),
            ('sum', '건수', '폭력', '검거'),
            ('sum', '건수', '폭력', '발생')],
           names=[None, None, '죄종', '발생검거'])

In [36]:
crime_station['sum','건수','강도','검거'].head()

구분
강남    26.0
강동    13.0
강북     4.0
강서    10.0
관악    10.0
Name: (sum, 건수, 강도, 검거), dtype: float64

In [37]:
crime_station.columns = crime_station.columns.droplevel([0,1])

In [38]:
crime_station.head()

죄종       강간           강도          살인           절도              폭력        
발생검거     검거     발생    검거    발생    검거   발생      검거      발생      검거      발생
구분                                                                       
강남    269.0  339.0  26.0  24.0   3.0  3.0  1129.0  2438.0  2096.0  2336.0
강동    152.0  160.0  13.0  14.0   5.0  4.0   902.0  1754.0  2201.0  2530.0
강북    159.0  217.0   4.0   5.0   6.0  7.0   672.0  1222.0  2482.0  2778.0
강서    239.0  275.0  10.0  10.0  10.0  9.0  1070.0  1952.0  2768.0  3204.0
관악    264.0  322.0  10.0  12.0   7.0  6.0   937.0  2103.0  2707.0  3235.0

In [39]:
crime_station.index

Index(['강남', '강동', '강북', '강서', '관악', '광진', '구로', '금천', '남대문', '노원', '도봉',
       '동대문', '동작', '마포', '방배', '서대문', '서부', '서초', '성동', '성북', '송파', '수서',
       '양천', '영등포', '용산', '은평', '종로', '종암', '중랑', '중부', '혜화'],
      dtype='object', name='구분')

- 현재 index는 경찰서 이름으로 되어있다.
- 경찰서 이름으로 구 이름을 알아내야 한다.

### 4. python 모듈설치

##### pip 명령
- python 의 공식 모듈 관리자
- pip list
- pip install module_name
- pip uninstall module_name
- mac(M1)

In [40]:
#pip list
get_ipython().system("pip list")

Package                   Version
------------------------- ------------
anyio                     3.5.0
argon2-cffi               21.3.0
argon2-cffi-bindings      21.2.0
asttokens                 2.0.5
async-lru                 2.0.4
attrs                     23.1.0
Babel                     2.11.0
backcall                  0.2.0
beautifulsoup4            4.12.2
bleach                    4.1.0
Bottleneck                1.3.5
Brotli                    1.0.9
certifi                   2023.11.17
cffi                      1.16.0
charset-normalizer        2.0.4
colorama                  0.4.6
comm                      0.1.2
contourpy                 1.0.5
cryptography              41.0.3
cycler                    0.11.0
debugpy                   1.6.7
decorator                 5.1.1
defusedxml                0.7.1
et-xmlfile                1.1.0
executing                 0.8.3
fastjsonschema            2.16.2
fonttools                 4.25.0
idna                      3.4
importlib-metadata

##### conda 명령
- conda list
- conda install module_name
- conda uninstall module_name
- conda install -c channel name module_name
 지정된 배포 채널에서 모듈 설치
- windows, mac(intel)


### 5. Google Maps API 설치

In [41]:
# 구글 계정
# AIzaSyCIFZq92_yAlZIH4NaJtt8A6YfxCoi21k0

In [42]:
# windows, mac(intel)
#API가 제대로 발급되었는지 확인하는 절차 (미니콘다 프롬프트)
#conda activate ds_study
#conda install -c conda-forge googlemaps

In [45]:
import googlemaps

In [46]:
gmaps_key = 'AIzaSyCIFZq92_yAlZIH4NaJtt8A6YfxCoi21k0'
gmaps = googlemaps.Client(key=gmaps_key)


In [47]:
gmaps.geocode('서울영등포경찰서' , language='ko')

[{'address_components': [{'long_name': '영등포구',
    'short_name': '영등포구',
    'types': ['political', 'sublocality', 'sublocality_level_1']},
   {'long_name': '서울특별시',
    'short_name': '서울특별시',
    'types': ['administrative_area_level_1', 'political']},
   {'long_name': '대한민국',
    'short_name': 'KR',
    'types': ['country', 'political']}],
  'formatted_address': '대한민국 서울특별시 영등포구',
  'geometry': {'bounds': {'northeast': {'lat': 37.556286, 'lng': 126.9498867},
    'southwest': {'lat': 37.4850011, 'lng': 126.8779285}},
   'location': {'lat': 37.5223245, 'lng': 126.9101692},
   'location_type': 'APPROXIMATE',
   'viewport': {'northeast': {'lat': 37.556286, 'lng': 126.9498867},
    'southwest': {'lat': 37.4850011, 'lng': 126.8779285}}},
  'partial_match': True,
  'place_id': 'ChIJk_FAvx2ffDURQAuECsjy0Cw',
  'types': ['political', 'sublocality', 'sublocality_level_1']}]

-------------

### Python 반복문
- 간단한 for문 예제

In [48]:
for n in [1,2,3,4]:
    print('Number is', n)

Number is 1
Number is 2
Number is 3
Number is 4


- 조금 복잡한 for문 예제

In [49]:
for n in range(0,10):
    print(n**2)


0
1
4
9
16
25
36
49
64
81


- 위 코드를 한 줄로: list comprehension

In [50]:
[n ** 2 for n in range(0,10)]

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

#### Pandas 에 잘 맞춰진 반복문용 명령 iterrows()
- pandas 데이터 프레임은 대부분 2차원
- 이럴 떄 for 문을 사용하면, n 번째라는 지정을 반복해서 가독률이 떨어짐
- pandas 데이터 프레임으로 반복문을 만들 때 iterrows() 옵션을 사용하면 편함
- 받을 떄, 인덱스와 내용으로 나누어 받는 것만 주의

### 6. Googla Maps 를 이용한 데이터 정리

In [51]:
import googlemaps

In [52]:
gmaps_key = 'AIzaSyCIFZq92_yAlZIH4NaJtt8A6YfxCoi21k0'
gmaps = googlemaps.Client(key=gmaps_key)


In [53]:
gmaps.geocode("서울영등포경찰서", language='ko') #단순 테스트 코드

[{'address_components': [{'long_name': '영등포구',
    'short_name': '영등포구',
    'types': ['political', 'sublocality', 'sublocality_level_1']},
   {'long_name': '서울특별시',
    'short_name': '서울특별시',
    'types': ['administrative_area_level_1', 'political']},
   {'long_name': '대한민국',
    'short_name': 'KR',
    'types': ['country', 'political']}],
  'formatted_address': '대한민국 서울특별시 영등포구',
  'geometry': {'bounds': {'northeast': {'lat': 37.556286, 'lng': 126.9498867},
    'southwest': {'lat': 37.4850011, 'lng': 126.8779285}},
   'location': {'lat': 37.5223245, 'lng': 126.9101692},
   'location_type': 'APPROXIMATE',
   'viewport': {'northeast': {'lat': 37.556286, 'lng': 126.9498867},
    'southwest': {'lat': 37.4850011, 'lng': 126.8779285}}},
  'partial_match': True,
  'place_id': 'ChIJk_FAvx2ffDURQAuECsjy0Cw',
  'types': ['political', 'sublocality', 'sublocality_level_1']}]

In [54]:
tmp = gmaps.geocode("서울영등포경찰서", language='ko')
len(tmp)

1

In [55]:
tmp[0]

{'address_components': [{'long_name': '영등포구',
   'short_name': '영등포구',
   'types': ['political', 'sublocality', 'sublocality_level_1']},
  {'long_name': '서울특별시',
   'short_name': '서울특별시',
   'types': ['administrative_area_level_1', 'political']},
  {'long_name': '대한민국',
   'short_name': 'KR',
   'types': ['country', 'political']}],
 'formatted_address': '대한민국 서울특별시 영등포구',
 'geometry': {'bounds': {'northeast': {'lat': 37.556286, 'lng': 126.9498867},
   'southwest': {'lat': 37.4850011, 'lng': 126.8779285}},
  'location': {'lat': 37.5223245, 'lng': 126.9101692},
  'location_type': 'APPROXIMATE',
  'viewport': {'northeast': {'lat': 37.556286, 'lng': 126.9498867},
   'southwest': {'lat': 37.4850011, 'lng': 126.8779285}}},
 'partial_match': True,
 'place_id': 'ChIJk_FAvx2ffDURQAuECsjy0Cw',
 'types': ['political', 'sublocality', 'sublocality_level_1']}

In [56]:
tmp[0].get("geometry")

{'bounds': {'northeast': {'lat': 37.556286, 'lng': 126.9498867},
  'southwest': {'lat': 37.4850011, 'lng': 126.8779285}},
 'location': {'lat': 37.5223245, 'lng': 126.9101692},
 'location_type': 'APPROXIMATE',
 'viewport': {'northeast': {'lat': 37.556286, 'lng': 126.9498867},
  'southwest': {'lat': 37.4850011, 'lng': 126.8779285}}}

In [57]:
tmp[0].get("geometry")['location']

{'lat': 37.5223245, 'lng': 126.9101692}

In [58]:
print(tmp[0].get("geometry")['location']['lat'])
print(tmp[0].get("geometry")['location']['lng'])

37.5223245
126.9101692


In [59]:
tmp[0].get("formatted_address")

'대한민국 서울특별시 영등포구'

In [60]:
# 영등포구 라는 단어만 얻어오기 위해 split() 사용

tmp[0].get("formatted_address").split()[2]

'영등포구'

In [61]:
crime_station.head()

죄종       강간           강도          살인           절도              폭력        
발생검거     검거     발생    검거    발생    검거   발생      검거      발생      검거      발생
구분                                                                       
강남    269.0  339.0  26.0  24.0   3.0  3.0  1129.0  2438.0  2096.0  2336.0
강동    152.0  160.0  13.0  14.0   5.0  4.0   902.0  1754.0  2201.0  2530.0
강북    159.0  217.0   4.0   5.0   6.0  7.0   672.0  1222.0  2482.0  2778.0
강서    239.0  275.0  10.0  10.0  10.0  9.0  1070.0  1952.0  2768.0  3204.0
관악    264.0  322.0  10.0  12.0   7.0  6.0   937.0  2103.0  2707.0  3235.0

In [62]:
crime_station.index

Index(['강남', '강동', '강북', '강서', '관악', '광진', '구로', '금천', '남대문', '노원', '도봉',
       '동대문', '동작', '마포', '방배', '서대문', '서부', '서초', '성동', '성북', '송파', '수서',
       '양천', '영등포', '용산', '은평', '종로', '종암', '중랑', '중부', '혜화'],
      dtype='object', name='구분')

- 구별 , lat, lng 컬럼 추가

In [63]:
crime_station['구별'] = np.nan # 구별컬럼 nan 값으로 추가
crime_station['lat'] = np.nan
crime_station['lng'] = np.nan

In [64]:
crime_station.head()

죄종       강간           강도          살인           절도              폭력          구별  \
발생검거     검거     발생    검거    발생    검거   발생      검거      발생      검거      발생       
구분                                                                              
강남    269.0  339.0  26.0  24.0   3.0  3.0  1129.0  2438.0  2096.0  2336.0 NaN   
강동    152.0  160.0  13.0  14.0   5.0  4.0   902.0  1754.0  2201.0  2530.0 NaN   
강북    159.0  217.0   4.0   5.0   6.0  7.0   672.0  1222.0  2482.0  2778.0 NaN   
강서    239.0  275.0  10.0  10.0  10.0  9.0  1070.0  1952.0  2768.0  3204.0 NaN   
관악    264.0  322.0  10.0  12.0   7.0  6.0   937.0  2103.0  2707.0  3235.0 NaN   

죄종   lat lng  
발생검거          
구분            
강남   NaN NaN  
강동   NaN NaN  
강북   NaN NaN  
강서   NaN NaN  
관악   NaN NaN

- 경찰서 이름에서 소속된 구이름을 얻기
- 구이름과 위도, 경도 정보를 저장할 준비
- 반복문을 이용해서 위 표의 nan 값을 모두 채워줍니다.

In [66]:
count = 0

for idx, rows in crime_station.iterrows():
    
    station_name = '서울' + str(idx) + '경찰서'
    tmp = gmaps.geocode(station_name, language='ko')
    
    if tmp:
    
        tmp[0].get('formatted_address')
        tmp_gu = tmp[0].get('formatted_address')
        
        lat = tmp[0].get("geometry")['location']['lat']
        lng = tmp[0].get("geometry")['location']['lng']
        
        crime_station.loc[idx, 'lat'] = lat #loc로 접근해서 idx 값으로 간다.
        crime_station.loc[idx, 'lng'] = lng
        crime_station.loc[idx, '구별'] = tmp_gu.split()[2]
    
    
    print(count)
    count += 1

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
